New package: dire

wdconinc opened the pull request:
DIRE (short for dipole resummation) a C++ program for all-order radiative corrections to scattering processes in high-energy particle collisions.

Link: http://dire.gitlab.io/

(Part of an effort to contribute spack packages from the Electron-Ion Collider repository to the broader NP and HEP community.)

wdconinc committed:
[dire] New package

tldahlgren reviewed (commented):
LGTM.  The current test failures are non-blockers.  Currently building a couple of versions to see results.

tldahlgren reviewed (changes requested):
I am able to build the latest (`2.004`) but none of the previously listed versions.  

The install process does not work for older versions of the package.  At a minimum, the process for versions `1.500` and `2.000` should take into account the fact that there is no `configure` script in the source directory.

Issues I'm seeing using `gcc@8.1.0` on a Linux machine at LLNL for `2.001` through `2.003` are build failures with `error: no matching function call` or `error: no matching operator` for multiple lines in `MergingHooks.cc` and, for `2.003`, two places in `History.cc`.

wdconinc committed:
[dire] Remove older versions (same physics and more included in 2.004)

Older versions do not compile correctly. New users should use 2.004,
not any of the older versions.

wdconinc commented:
Rather than resolve the compilation issues for versions that are unlikely to be used in practice, I simply removed the older versions.

tldahlgren reviewed (approved):
LGTM.  The addition of the when to the `pythia8` dependency is technically not necessary if you remove all of the previous versions but, IMHO, it is good to be explicit.

tldahlgren commented:
I will merge this PR once the required `macos tests / build` pass.  There is currently a `brew` update problem affecting multiple PRs.

adamjstewart commented on var/spack/repos/builtin/packages/dire/package.py:
```diff
@@ -0,0 +1,37 @@
+from spack import *
+
+
+class Dire(Package):
```
You should extend `AutotoolsPackage` here

wdconinc replied:
I considered it, but it is not in fact an autotools package.
```
14:53:03 wdconinc@herakles /tmp/dire/DIRE-2.004 $ ls
bin  configure  include  INSTALL  main  Makefile  Makefile.inc  README  share  src  VERSION
```

adamjstewart commented on var/spack/repos/builtin/packages/dire/package.py:
```diff
@@ -0,0 +1,37 @@
+    depends_on('boost')
+    depends_on('lhapdf')
+    depends_on('hepmc')
+    depends_on('pythia8@8226:', when='@2.004:')
```
```suggestion
    depends_on('pythia8@8226:')
```

adamjstewart commented on var/spack/repos/builtin/packages/dire/package.py:
```diff
@@ -0,0 +1,37 @@
+        configure(*configure_args)
+        make()
+        # Open bug: https://gitlab.com/wdconinc/direforpythia/-/merge_requests/1
+        filter_file('-Wl,-rpath ', '-Wl,-rpath,', 'bin/dire-config')
```
I wonder how this affects other compilers like NAG who use a different RPATH flag

wdconinc replied:
Unable to test at this point. I don't have access to a system with an existing NAG compiler. But I've requested a trial license, will test and report.

wdconinc replied:
I think I understand better now. NAG would require `-Wl,-Wl,,-rpath,` to pass this to the linker correctly. I'm not sure it is in scope to try to resolve this here though. dire-config is just a helper script for user code to allow e.g. ``gcc `dire-config --libs` source.cc``. dire-config does not have any way of knowing which compiler is used. Thoughts?

adamjstewart replied:
If it's for C/C++ only, then I think the rpath flags would be the same since NAG is only a Fortran compiler. If you want to be careful, you can use `self.compiler.cc_rpath_arg`, but all the C/C++ compilers we currently support use the default flag.

wdconinc replied:
Implemented, tested, verified that the correct rpath lines appear in dire-config.

wdconinc committed:
[dire] depend_on pythia8@8226: for all supported versions

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

wdconinc committed:
[dire] use compiler.cc_rpath_arg instead of '-Wl,-rpath,'

wdconinc committed:
[dire] flake8: long line to multiline
